File: webnetlab/frontend/src/types/topology.ts
export interface TopologyNode {
  id: number;
  name: string;
  type: string;
  ip_address: string;
  status: string;
  interfaces: string[];
}

export interface TopologyLink {
  id: number;
  src_device_id: number;
  src_interface: string;
  dst_device_id: number;
  dst_interface: string;
  docker_network_id: string | null;
}

export interface TopologyData {
  nodes: TopologyNode[];
  links: TopologyLink[];
}

File: webnetlab/backend/app/schemas/topology.py
from datetime import datetime

from pydantic import BaseModel


class TopologyLinkCreate(BaseModel):
    src_device_id: int
    src_interface: str      # e.g. "GigabitEthernet0/0"
    dst_device_id: int
    dst_interface: str


class TopologyLinkResponse(BaseModel):
    id: int
    src_device_id: int
    src_interface: str
    dst_device_id: int
    dst_interface: str
    docker_network_id: str | None
    created_at: datetime
    model_config = {"from_attributes": True}


class TopologyNodeResponse(BaseModel):
    id: int
    name: str
    type: str
    ip_address: str
    status: str
    interfaces: list[str]


class TopologyResponse(BaseModel):
    nodes: list[TopologyNodeResponse]
    links: list[TopologyLinkResponse]
